Add troubleshooting scenarios for kubevirt disconected (HCP)

xenolinux opened the pull request:
https://issues.redhat.com/browse/ACM-11225

xenolinux committed:
Add troubleshooting scenarios for kubevirt disconected (HCP)

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,54 @@
+For hosted control planes on {ocp-virt} in a disconnected environment, oc-mirror fails to automatically mirror the Red Hat Enterprise Linux CoreOS (RHCOS) image.
+
+[#symptom-mirror-rhcos-image]
+== Symptom: Mirroring of the RHCOS image fails
```
I'd suggest to rephrase this: the mirroring of the RHCOS boot image for Kubevirt is not really failing but it's not going to be attempted at all by oc-mirror.
This means that `oc-mirror` process will successfully complete with no errors but later on the VM will fail to boot since their OS is not there.

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,54 @@
+[source,bash]
+----
+oc image extract --file /release-manifests/<installer_coreos_bootimages.yaml> <payload_image> --confirm
+----
```
`<installer_coreos_bootimages.yaml>` is `always 0000_50_installer_coreos-bootimages.yaml`, no need to have the cluster admin looking for it.

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,54 @@
++
+[source,bash]
+----
+cat <installer_coreos_bootimages.yaml> | yq -r .data.stream | jq -r '.architectures.x86_64.images.kubevirt."digest-ref"'
```
same

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,54 @@
++
+[source,bash]
+----
+oc image mirror <rhcos_image> <mirror_registry_name>/<local_images>/<rhcos_image>%@sha256*##*/
```
`<local_images>` can be a fixed pattern but it has to be used also in the ImageDigestMirrorSet.
`%@sha256*##*` was intended to skip whatever after `@sha256` (including `@sha256` itself) in bash variable expansion. It will never work like this.

tiraboschi replied:
Let me provide a complete example here:
```bash
# virthost.ostest.test.metalkube.org:5000/localimages/local-release-image@sha256:a21f3f3f3c77414b0c611c7b3e8fe301dfae0eb6a765a4c4f8305c8acee9258d is the OpenShift release image as mirrored to the internal registry
# virthost.ostest.test.metalkube.org:5000 is the address of the internal registry
# localimages is a path that the user can potentially replace with a custom string
oc image extract virthost.ostest.test.metalkube.org:5000/localimages/local-release-image@sha256:a21f3f3f3c77414b0c611c7b3e8fe301dfae0eb6a765a4c4f8305c8acee9258d --path /release-manifests/0000_50_installer_coreos-bootimages.yaml --confirm
# /release-manifests/0000_50_installer_coreos-bootimages.yaml is a fixed path that is never going to change
oc image extract virthost.ostest.test.metalkube.org:5000/localimages/local-release-image@sha256:a21f3f3f3c77414b0c611c7b3e8fe301dfae0eb6a765a4c4f8305c8acee9258d --file /release-manifests/0000_50_installer_coreos-bootimages.yaml --confirm
# this will return the pull URL of the image we want to mirror, in my example quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94
cat 0000_50_installer_coreos-bootimages.yaml | yq -r .data.stream | jq -r '.architectures.x86_64.images.kubevirt."digest-ref"'
# quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94 should be splitted intro fragments to compose a command like:
oc image mirror quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94 virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev
# where quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94 is the pull URL as extracted from the file
# virthost.ostest.test.metalkube.org:5000 is the internal registry
# localimages is a string for the path that the user can customize but it should be consistent across the whole process
# ocp-v4.0-art-dev is the name of the image as the last piece of the pull URL before @sha256, this should match what extracted from the file
# finally the cluster admin should apply a matching ImageDigestMirrorSet:
oc apply -f - <<EOF
apiVersion: config.openshift.io/v1
kind: ImageDigestMirrorSet
metadata:
  name: rhcos-boot-kubevirt
spec:
  repositoryDigestMirrors:
    - mirrors:
        - quay.io/openshift-release-dev/ocp-v4.0-art-dev
      source: virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev
EOF
```

tiraboschi replied:
I still think that `oc image mirror <rhcos_image> <mirror_registry_name>/<local_images>/<rhcos_image>%@sha256*##*/` line here should be fixed.
The expected command is something like:
`oc image mirror quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94 virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev`
where `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94` is  `<rhcos_image>`  as referenced (including the sha digest) in `0000_50_installer_coreos-bootimages.yaml` while `virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev` is the new name on the internal registry and it should not include the digest.

tiraboschi commented on clusters/support_troubleshooting/trouble_set_import_policy.adoc:
```diff
@@ -0,0 +1,30 @@
+[#trouble-set-import-policy]
```
https://issues.redhat.com/browse/OCPBUGS-31446 has been fixed and verified so we can completely drop this workaround

xenolinux replied:
@tiraboschi Updated the PR -- can you PTAL

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+
+The oc-mirror plugin does not mirror the {op-system-first} image from the release payload to the internal registry.
+
+[#identifying-mirror-rhcos-image]
```
shouldn't the title match the anchor? 

So the anchor could be something like: 

```suggestion
[#resolving-mirror-rhcos-image]
```

swopebe replied:
The format for these files should always be the same. If there is no resolving or work around, then it is not t-shooting. If identifying the problem is troubleshooting then it is resolving.

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+For hosted control planes on {ocp-virt} in a disconnected environment, oc-mirror fails to automatically mirror the Red Hat Enterprise Linux CoreOS (RHCOS) image to the internal registry.
+
+[#symptom-mirror-rhcos-image]
+== Symptom: oc-mirror fails to attempt the RHCOS image mirroring
```
We're trying to give shorter titles. If "to attempt" is needed, then keep it. That is, if the failure only happens because of an attempt. Otherwise, consider shortening to something like: 

```suggestion
== Symptom: oc-mirror fails the RHCOS image mirroring
```

xenolinux replied:
https://github.com/stolostron/rhacm-docs/pull/6519#discussion_r1634843840 I will go with keeping the title as is, per SME's comment

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+The oc-mirror plugin does not mirror the {op-system-first} image from the release payload to the internal registry.
+
+[#identifying-mirror-rhcos-image]
+== Resolving the problem: Manually mirror the RHCOS image to the internal registry
```
Again with an optional change to the heading since we're trying to make them shorter: 

```suggestion
== Resolution: Manually mirror the RHCOS image
```

oafischer replied:
**Resolving the problem** is correct for troubleshooting.

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+[#identifying-mirror-rhcos-image]
+== Resolving the problem: Manually mirror the RHCOS image to the internal registry
+
+You need to manually mirror the RHCOS image by using the payload image to the internal registry. Complete the following steps:
```
To make it clear this action is to resolve something, consider rephrasing to the following: 
```suggestion
To resolve this issue, manually mirror the RHCOS image by using the payload image to the internal registry. Complete the following steps:
```

swopebe replied:
I don't agree with this one as much because the header is enough, so those words are extra. It's not wrong either way but I want to emphasize that the header is the important aspect here. The header is what is read in an accessibility situation. That is why the anchor and the title need to be similar.

This should be **resolving** in the anchor. @xenolinux

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+oc get clusterversion version -ojsonpath='{.status.desired.image}'
+----
+
+. Extract the `0000_50_installer_coreos-bootimages.yaml` file that contains boot images and the payload image. Replace `<payload_image>` with the name of your payload image. Run the following command:
```
Separate steps? And what does running the following command do? Correct me if I'm mistaken, but see the possible revisions: 


```suggestion
. Extract the `0000_50_installer_coreos-bootimages.yaml` file that contains boot images and the payload image. 
. Replace `<payload_image>` with the name of your payload image by running the following command:
```

xenolinux replied:
The command extracts the payload image into the `0000_50_installer_coreos-bootimages.yaml` file. Updated the step as such.

xenolinux replied:
The command extracts the payload image into the `0000_50_installer_coreos-bootimages.yaml` file. Updated the step as such.

xenolinux replied:
IMO, two separate steps are not needed here because users need to enter payload image in the extract command. Also the instruction is updated.

swopebe replied:
replacing an item in a command is part of the step, you cannot run the command until you replace the item.

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+cat 0000_50_installer_coreos-bootimages.yaml | yq -r .data.stream | jq -r '.architectures.x86_64.images.kubevirt."digest-ref"'
+----
+
+. Mirror the RHCOS image to your internal registry. Replace `<rhcos_image>` with your RHCOS image for example, `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94`. Replace `<internal_registry>` with the name of your internal registry, for example, `virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev`. Run the fo ...
```
```suggestion
. Mirror the RHCOS image to your internal registry. Replace `<rhcos_image>` with your RHCOS image for example, `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9643ead36b1c026be664c9c65c11433c6cdf71bfd93ba229141d134a4a6dd94`. Replace `<internal_registry>` with the name of your internal registry, for example: `virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev`. Run the following command:
```

xenolinux replied:
keeping this as is.

swopebe replied:
so the command here should be first, no? Then we tell the user what to replace?

Run the following command to _____. Replace x with y.

This line: Replace `<internal_registry>` with the name of your internal registry, for example: 

Consider clarity for a screen reader:

Replace `<internal_registry>` with the name of your internal registry, as displayed in the following example: 

or

Replace `<internal_registry>` with the name of your internal registry. For example, see the following internal registry.

(it's not wrong, it just doesn't tell the user what follows the colon)

xenolinux replied:
Per slack discussion - keeping this as is because we need a colon in the end of this command.

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+----
+
++
+<1> Specify your RHCOS image without its digest, for example, `quay.io/openshift-release-dev/ocp-v4.0-art-dev`.
```
```suggestion
<1> Specify your RHCOS image without its digest, for example: `quay.io/openshift-release-dev/ocp-v4.0-art-dev`.
```

xenolinux replied:
keeping this as is.

jc-berger commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+
++
+<1> Specify your RHCOS image without its digest, for example, `quay.io/openshift-release-dev/ocp-v4.0-art-dev`.
+<2> Specify the name of your internal registry, for example, `virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev`.
```
```suggestion
<2> Specify the name of your internal registry, for example: `virthost.ostest.test.metalkube.org:5000/localimages/ocp-v4.0-art-dev`.
```

xenolinux replied:
I see more instances of `for example,` in the repo. keeping this as is.

swopebe replied:
Yeah, I will chat about that in the meeting. It's not wrong, it's just not the most accessible in either of these examples.

oafischer commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+[#trouble-mirror-rhcos-image]
+= Troubleshooting the RHCOS image mirroring
+
+For hosted control planes on {ocp-virt} in a disconnected environment, oc-mirror fails to automatically mirror the Red Hat Enterprise Linux CoreOS (RHCOS) image to the internal registry.
```
```suggestion
For hosted control planes on {ocp-virt} in a disconnected environment, `oc-mirror` fails to automatically mirror the Red Hat Enterprise Linux CoreOS (RHCOS) image to the internal registry.
```

We seem to use backticks for it throughout the doc so I recommend doing the same here

swopebe replied:
Yes, this should be coded. If this was translated, it would not make sense in another language. The tics tells the translator that it is part of the code.

oafischer commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+For hosted control planes on {ocp-virt} in a disconnected environment, oc-mirror fails to automatically mirror the Red Hat Enterprise Linux CoreOS (RHCOS) image to the internal registry.
+
+[#symptom-mirror-rhcos-image]
+== Symptom: oc-mirror fails to attempt the RHCOS image mirroring
```
```suggestion
== Symptom: _oc-mirror_ fails to attempt the RHCOS image mirroring
```

oafischer commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+[#symptom-mirror-rhcos-image]
+== Symptom: oc-mirror fails to attempt the RHCOS image mirroring
+
+The oc-mirror plugin does not mirror the {op-system-first} image from the release payload to the internal registry.
```
```suggestion
The `oc-mirror` plugin does not mirror the {op-system-first} image from the release payload to the internal registry.
```

xenolinux replied:
per slack discussion, keeping as is.

swopebe replied:
This should be coded.

oafischer commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+The oc-mirror plugin does not mirror the {op-system-first} image from the release payload to the internal registry.
+
+[#resolving-mirror-rhcos-image]
+== Resolution: Manually mirror the RHCOS image to the internal registry
```
You are missing the `Identifying the problem` section. Take a look at one of our existing troubleshoot topics and just copy the format: https://github.com/stolostron/rhacm-docs/blob/2.11_stage/troubleshooting/trouble_already_exists.adoc

Note that the title has to be the same for each of the 3 sections.

swopebe replied:
If we don't have one, resolution is more important. Not all the files have them, we will chat about this.

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+oc get clusterversion version -ojsonpath='{.status.desired.image}'
+----
+
+. Extract the payload image into the `0000_50_installer_coreos-bootimages.yaml` file that contains boot images. Replace `<payload_image>` with the name of your payload image. Run the following command:
```
this is now misleading: this is not going to extract the whole `<payload_image>` into a file, but just a file named `/release-manifests/0000_50_installer_coreos-bootimages.yaml` from that image to local filesystem.
`0000_50_installer_coreos-bootimages.yaml` then contains the reference to the RHCOS boot image for KubeVirt.

xenolinux replied:
Ok. Reverted the instruction.

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+[#resolving-mirror-rhcos-image]
+== Resolution: Manually mirror the RHCOS image to the internal registry
+
+To resolve this issue, manually mirror the RHCOS image by using the payload image to the internal registry. Complete the following steps:
```
The goal is to know the address of the missing image (the RHCOS boot image for KubeVirt that got skipped in the oc-mirror process) so that we cluster admin can manually/directly mirror it.
In order to do that, he is going to extract and inspect a specific file from the release payload image that got previously mirrored with oc-mirror.

xenolinux commented:
@tiraboschi Can you provide any error msg, I need an error in the **identifying the problem** section. OR a tiny context of an error would help too

tiraboschi commented:
> @tiraboschi Can you provide any error msg, I need an error in the **identifying the problem** section. OR a tiny context of an error would help too

Unfortunately `oc-mirror` is completely ignoring the existence of the RHCOS boot image so the oc-mirror process will successfully complete with no error.
Only when you will attempt to create your first guest cluster, the Kubevirt VM will not boot since the boot image is not there in the internal registry but the error is not self-explicative: the cdi-importer pod (the pod that is writing the content of the RHCOS boot image to the PVC used as the disk of the KubeVirt VM) will be stuck forever. Checking its log you will be able to see looping ImagePullBackOff errors for RHCOS boot image but the name of the image is even not self-explanatory.

xenolinux commented:
> > @tiraboschi Can you provide any error msg, I need an error in the **identifying the problem** section. OR a tiny context of an error would help too
> 
> Unfortunately `oc-mirror` is completely ignoring the existence of the RHCOS boot image so the oc-mirror process will successfully complete with no error. Only when you will attempt to create your first guest cluster, the Kubevirt VM will not boot since the boot image is not there in the internal registry but the error is not self-explicative: the cdi-importer pod (the pod that is writing the content of the RHCOS boot image to the PVC used as the disk of the KubeVirt VM) will be stuck forever. Checking its log you will be able to see looping ImagePullBackOff errors for RHCOS boot image but the name of the image is even not self-explanatory.

I added this info in the symptom section, can you PTAL? And I replaced "guest cluster" with "hosted cluster"

tiraboschi commented on clusters/support_troubleshooting/trouble_mirror_rhcos_image.adoc:
```diff
@@ -0,0 +1,82 @@
+oc get clusterversion version -ojsonpath='{.status.desired.image}'
+----
+
+. Extract the payload image and the `0000_50_installer_coreos-bootimages.yaml` file that contains boot images. Replace `<payload_image>` with the name of your payload image. Run the following command:
```
```suggestion
. Extract file `0000_50_installer_coreos-bootimages.yaml` from the release payload image you intend to use on the hosted cluster. That file contains the list of boot images. Replace `<payload_image>` with the name of your payload image. Run the following command:
```

xenolinux replied:
Done

tiraboschi commented:
/lgtm
